hoomd/mpcd/pytest/test_update.py

```diff
@@ -0,0 +1,184 @@
+        snap.particles.N = 0
+        snap.particles.types = ["A"]
+        snap.mpcd.types = ["A"]
+        snap.mpcd.mass = 1.0
+    return snap
+
+
+@pytest.fixture
+def flow():
+    flow = hoomd.mpcd.update.ReverseNonequilibriumShearFlow(
+        trigger=hoomd.trigger.Periodic(1),
+        num_swaps=1,
+        slab_width=1.0,
+        slab_separation=10.0,
+        target_momentum=2.0)
+    return flow
+
+
+class TestReverseNonequilibriumShearFlow:
+
+    def test_create(self, simulation_factory, snap, flow):
+        # before attachment
+        assert flow.trigger == hoomd.trigger.Periodic(1)
+        assert flow.num_swaps == 1
+        assert flow.slab_width == 1.0
+        assert flow.slab_separation == 10.0
+        assert flow.target_momentum == 2.0
+        with pytest.raises(hoomd.error.DataAccessError):
+            assert flow.summed_exchanged_momentum == 0.0
+
+        sim = simulation_factory(snap)
+        sim.operations.integrator = hoomd.mpcd.Integrator(dt=0.09)
+        sim.operations.add(flow)
+        sim.run(0)
+
+        # after attachment
+        assert flow.trigger == hoomd.trigger.Periodic(1)
+        assert flow.num_swaps == 1
+        assert flow.slab_width == 1.0
+        assert flow.slab_separation == 10.0
+        assert flow.target_momentum == 2.0
+        assert flow.summed_exchanged_momentum == 0.0
+
+    def test_pickling(self, simulation_factory, snap, flow):
+        pickling_check(flow)
+
+        flow.trigger = hoomd.trigger.Periodic(1)
+        flow.num_swaps = 5
+        pickling_check(flow)
+
+        sim = simulation_factory(snap)
+        sim.operations.integrator = hoomd.mpcd.Integrator(dt=0.09)
+        sim.operations.add(flow)
+        sim.run(1)
+        pickling_check(flow)
+
+    # particle locations in/out of the slabs; last 4 are at the boundaries
+    @pytest.mark.parametrize(
+        "v0, v1, y0, y1, swap",
+        [[-2, 2, 5, -5, True], [2, -2, 5, -5, False], [2, 2, 5, -5, False],
+         [-2, -2, 5, -5, False], [-2, 2, 3, -3, False], [-2, 2, 5, -3, False],
+         [-2, 2, 3, -5, False], [-2, 2, 4.5, -5.5, True],
+         [-2, 2, 4.5, -4.5, False], [-2, 2, 5.5, -5.5, False],
+         [-2, 2, 5.5, -4.5, False]])
+    def test_particle_swapping(self, simulation_factory, snap, flow, v0, v1, y0,
+                               y1, swap):
+        snap.mpcd.N = 2
+        snap.mpcd.position[0] = [1.0, y0, 2.0]
+        snap.mpcd.position[1] = [1.0, y1, 2.0]
+        snap.mpcd.velocity[0] = [v0, 4.0, 5.0]
+        snap.mpcd.velocity[1] = [v1, 1.0, 3.0]
+
+        sim = simulation_factory(snap)
+        sim.operations.integrator = hoomd.mpcd.Integrator(dt=0.09)
+        sim.operations.add(flow)
+        sim.run(1)
+        snap2 = sim.state.get_snapshot()
+
+        assert np.allclose(snap2.mpcd.velocity[0],
+                           [v1 if swap else v0, 4.0, 5.0])
+        assert np.allclose(snap2.mpcd.velocity[1],
+                           [v0 if swap else v1, 1.0, 3.0])
```

Comment:
Two things: (1) it's fine to recycle the `snap` variable name rather than keep numbering it, and (2) we should wrap these in a rank check so that these tests continue to work when we enable MPI.
```suggestion

        snap = sim.state.get_snapshot()
        if snap.communicator.rank == 0:
            assert np.allclose(snap.mpcd.velocity[0],
                               [v1 if swap else v0, 4.0, 5.0])
            assert np.allclose(snap.mpcd.velocity[1],
                               [v0 if swap else v1, 1.0, 3.0])
```
Please do similarly for the other places where you take a snapshot.